Module Enable Mode + MultiType Module Installation

filmakarov opened the pull request:
### Multi Type Module Installation
- Provide `0x00` as ModuleTypeId
- `abi.encode(uint256[] typeIds, bytes[] initDatas)` as module init data

### Enable Mode
- New structure for nonce: 
```
[3 bytes empty][1 bytes validation mode][20 bytes validator][8 bytes nonce]
```
- `0x00` for regular validation, `0x01` for Module Enable Mode
- Module Enable Mode data should be prepended to the userOp.signature
```
enableModeData = abi.encodePacked(
            moduleTypeId,
            bytes4(uint32(moduleInitData.length)),
            moduleInitData,
            bytes4(uint32(enableModeSig.length)),
            enableModeSig
        );
```
- `enableModeSig` should be issued by a current SA's owner/signer and it should be prepended with bytes20 address of the validator that is going to validate it.
- to obtain enableModeSig a proper EIP712 hash should be signed.

livingrockrises commented:
you do need to commit .yarn/plugins and release folder etc?

livingrockrises commented on contracts/Nexus.sol:
```diff
@@ -23,8 +22,16 @@ import { BaseAccount } from "./base/BaseAccount.sol";
+    VALIDATION_FAILED 
+} from "./types/Constants.sol";
 import { ModeLib, ExecutionMode, ExecType, CallType, CALLTYPE_BATCH, CALLTYPE_SINGLE, EXECTYPE_DEFAULT, EXECTYPE_TRY } from "./lib/ModeLib.sol";
+import { NonceLib } from "./lib/NonceLib.sol";
```
we require nonce lib for packing and unpacking nonce for validation?

filmakarov replied:
right
just unpacking currently

livingrockrises commented on contracts/Nexus.sol:
```diff
@@ -13,7 +13,6 @@ pragma solidity ^0.8.26;
 // Learn more at https://biconomy.io. To report security issues, please contact us at: security@biconomy.io
 
 import { UUPSUpgradeable } from "solady/src/utils/UUPSUpgradeable.sol";
-import { EIP712 } from "solady/src/utils/EIP712.sol";
```
why get rid of EIP712?

filmakarov replied:
it is in Module Manager now => inherited

livingrockrises commented on contracts/Nexus.sol:
```diff
@@ -72,20 +80,24 @@ contract Nexus is INexus, EIP712, BaseAccount, ExecutionHelper, ModuleManager, U
+        address validator = op.nonce.getValidator();
+        
+        if (op.nonce.isModuleEnableMode()) {
+            userOp.signature = _enableMode(validator, op.signature);
```
can be a better name is purpose is to retrieve signature in case of module enable mode.

livingrockrises replied:
I see it will also install the validator.. but it will verify that deed through only nonce?

livingrockrises replied:
just looking into the implementation now

filmakarov replied:
no, the purpose is to execute the enable flow, signature is just returned to save gas and not parse it again

livingrockrises commented on contracts/Nexus.sol:
```diff
@@ -72,20 +80,24 @@ contract Nexus is INexus, EIP712, BaseAccount, ExecutionHelper, ModuleManager, U
-        uint256 userOpnonce = userOp.nonce;
-        assembly {
-            validator := shr(96, userOpnonce)
+        PackedUserOperation memory userOp = op;
```
need to copy in memory first?

filmakarov replied:
yes, to change the signature later
moved it to if branch. looks less elegant but copies only when it s needed

livingrockrises commented on contracts/base/ModuleManager.sol:
```diff
@@ -31,7 +41,7 @@ import { IModuleManagerEventsAndErrors } from "../interfaces/base/IModuleManager
 /// @author @zeroknots | Rhinestone.wtf | zeroknots.eth
 /// Special thanks to the Solady team for foundational contributions: https://github.com/Vectorized/solady
-contract ModuleManager is Storage, Receiver, IModuleManagerEventsAndErrors {
+abstract contract ModuleManager is Storage, Receiver, EIP712, IModuleManagerEventsAndErrors {
```
any specific motivation for moving EIP712 here?

livingrockrises replied:
ok

Aboudjem commented on contracts/Nexus.sol:
```diff
@@ -267,6 +262,7 @@ contract Nexus is INexus, EIP712, BaseAccount, ExecutionHelper, ModuleManager, U
         else if (moduleTypeId == MODULE_TYPE_EXECUTOR) return true;
         else if (moduleTypeId == MODULE_TYPE_FALLBACK) return true;
         else if (moduleTypeId == MODULE_TYPE_HOOK) return true;
+        else if (moduleTypeId == MULTITYPE_MODULE) return true;
```
could we rename it MODULE_TYPE_MULTI instead for consistency ?

filmakarov replied:
done

livingrockrises commented on contracts/Nexus.sol:
```diff
@@ -546,21 +542,6 @@ contract Nexus is INexus, EIP712, BaseAccount, ExecutionHelper, ModuleManager, U
-    /// @param module The module address.
-    /// @param additionalContext Additional context for checking installation.
-    /// @return True if the module is installed, false otherwise.
-    function _isModuleInstalled(uint256 moduleTypeId, address module, bytes calldata additionalContext) private view returns (bool) {
```
@Aboudjem remind me again why was this kept in implementation contract?

Aboudjem replied:
no particular reason, I think we moved it while we were using the `withHook` modifier, but not really sure

livingrockrises commented on contracts/base/ModuleManager.sol:
```diff
@@ -246,6 +309,113 @@ contract ModuleManager is Storage, Receiver, IModuleManagerEventsAndErrors {
+    /// @param digest signed digest.
+    /// @param sig Signature.
+    function _checkEnableModeSignature(bytes32 digest, bytes calldata sig) internal view {
+        address enableModeSigValidator = address(bytes20(sig[0:20]));
```
this could be any of the installed validators?

filmakarov replied:
yes

livingrockrises commented on contracts/lib/local/LocalCallDataParserLib.sol:
```diff
@@ -0,0 +1,55 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.26;
+
+library LocalCallDataParserLib {
```
why 'Local' ?

filmakarov replied:
Because it is not universal. And most probably not reusable. It is just for the sake of code readability.

livingrockrises commented on contracts/mocks/MockMultiModule.sol:
```diff
@@ -0,0 +1,68 @@
+    function onInstall(bytes calldata data) external override {
+        if (data.length >= 0x21) {
+            uint256 moduleTypeId = uint256(uint8(bytes1(data[:1])));
+            configs[moduleTypeId][msg.sender] = bytes32(data[1:33]);
```
didn't get this logic

filmakarov replied:
just storing the config data based on module type. it is just a mock though

livingrockrises commented on contracts/utils/Bootstrap.sol:
```diff
@@ -121,4 +121,10 @@ contract Bootstrap is ModuleManager {
     }
+
+    /// @dev EIP712 domain name and version.
+    function _domainNameAndVersion() internal pure override returns (string memory name, string memory version) {
```
this is for completeness?

Aboudjem replied:
shouldn't be public/external?

filmakarov replied:
it is for completeness, yes.

no, it shouldn't as it is used in other public method

livingrockrises commented on test/foundry/unit/concrete/erc4337account/TestERC4337Account_OnlyEntryPoint.t.sol:
```diff
@@ -20,21 +21,37 @@ contract TestERC4337Account_OnlyEntryPoint is NexusTest_Base {
     function test_RevertIf_UserOpFromNonEntryPoint() public {
         PackedUserOperation[] memory userOps = new PackedUserOperation[](1);
-        userOps[0] = buildPackedUserOp(userAddress, getNonce(address(BOB_ACCOUNT), address(VALIDATOR_MODULE)));
+        userOps[0] = buildPackedUserOp(userAddress, getNonce(address(BOB_ACCOUNT), MODE_VALIDATION, address(VALIDATOR_MODULE)));
```
what kind of other modes could be there?

filmakarov replied:
idk yet, maybe MODE_DEFAULT if decide to introduce at some point

livingrockrises reviewed (commented):
review i

Aboudjem commented on codecov.yml:
what it is doing ? :)

filmakarov replied:
if you mean the string I added, it adds an exception not to account for libraries that consist almost completely of assembly blocks in code coverage

Aboudjem commented on test/foundry/fork/base/BaseSettings.t.sol:
```diff
@@ -8,7 +8,9 @@ import "../../utils/NexusTest_Base.t.sol";
     address public constant USDC_ADDRESS = 0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913;
-    string public constant DEFAULT_BASE_RPC_URL = "https://base.llamarpc.com";
+    string public constant DEFAULT_BASE_RPC_URL = "https://mainnet.base.org";
+    //string public constant DEFAULT_BASE_RPC_URL = "https://base.llamarpc.com";
+    //string public constant DEFAULT_BASE_RPC_URL = "https://developer-access-mainnet.base.org";
```
```suggestion
```

Aboudjem commented on test/foundry/fork/arbitrum/ArbitrumSettings.t.sol:
```diff
@@ -11,7 +11,10 @@ contract ArbitrumSettings is NexusTest_Base {
     address public constant USDC_ADDRESS = 0xFF970A61A04b1cA14834A43f5dE4533eBDDB5CC8;
-    string public constant DEFAULT_ARBITRUM_RPC_URL = "https://arbitrum-one-archive.allthatnode.com";
+    string public constant DEFAULT_ARBITRUM_RPC_URL = "https://public.stackup.sh/api/v1/node/arbitrum-one";
+    //string public constant DEFAULT_ARBITRUM_RPC_URL = "https://arb-pokt.nodies.app";
+    //string public constant DEFAULT_ARBITRUM_RPC_URL = "https://rpc.ankr.com/arbitrum";
+    //string public constant DEFAULT_ARBITRUM_RPC_URL = "https://arbitrum-one-rpc.publicnode.com";
```
```suggestion
```
Did you add them on purpose?  or just forgot to remove ?

filmakarov replied:
yeah, I added them on purpose as backup RPC URLS
as for me some tests were not passing because url unavailable. so can easily switch between them

Aboudjem commented on contracts/lib/local/LocalCallDataParserLib.sol:
lgtmp, would be great to add natspec here

Aboudjem commented on contracts/lib/NonceLib.sol:
```diff
@@ -0,0 +1,21 @@
+        assembly {
+            let vmode := shr(248, shl(24, nonce))
+            res := eq(shl(248, vmode), MODE_MODULE_ENABLE)
+            v := vmode
```
I'm not understanding why the `v := vmode` as only `res` is returned

filmakarov replied:
good catch, thank you! it's testing artifact

Aboudjem commented on contracts/lib/NonceLib.sol:
could you please add natspec here please

Aboudjem commented on contracts/lib/NonceLib.sol:
```diff
@@ -0,0 +1,21 @@
+        bytes32 v;
+        assembly {
+            let vmode := shr(248, shl(24, nonce))
+            res := eq(shl(248, vmode), MODE_MODULE_ENABLE)
```
would be great to add some inline comment `//` to briefly explain what we're trying to achieve there

Aboudjem commented on contracts/lib/NonceLib.sol:
```diff
@@ -0,0 +1,21 @@
+library NonceLib {
+    function getValidator(uint256 nonce) internal pure returns (address validator) {
+        assembly {
+            validator := shr(96, shl(32, nonce))
```
would be great to add some inline comment // to briefly explain what we're trying to achieve there

filmakarov replied:
added the nonce struct => the assembly code is easier to understand

Aboudjem commented on contracts/base/ModuleManager.sol:
```diff
@@ -246,6 +309,113 @@ contract ModuleManager is Storage, Receiver, IModuleManagerEventsAndErrors {
+        for (uint256 i; i < length; i++) {
+            uint256 theType = types[i];
+
+            /*´:°•.°+.*•´.*:˚.°*.˚•´.°:°•.°•.*•´.*:˚.°*.˚•´.°:°•.°+.*•´.*:*/
+            /*                      INSTALL VALIDATORS                    */
+            /*.•°:°.´+˚.*°.˚:*.´•*.+°.•°:´*.´•*.•°.•°:°.´:•˚°.*°.˚:*.´+°.•*/
```
this is beautiful

filmakarov replied:
Took it from zeroknots' contracts :)

Aboudjem commented on contracts/base/ModuleManager.sol:
```diff
@@ -139,10 +151,57 @@ contract ModuleManager is Storage, Receiver, IModuleManagerEventsAndErrors {
+    /// - 1 for Validator
+    /// - 2 for Executor
+    /// - 3 for Fallback
+    /// - 4 for Hook
```
Would it be beneficial to include multitype here as well?

filmakarov replied:
done

Aboudjem commented on contracts/Nexus.sol:
```diff
@@ -72,22 +80,25 @@ contract Nexus is INexus, EIP712, BaseAccount, ExecutionHelper, ModuleManager, U
-        // bubble up the return value of the validator module
-        validationData = IValidator(validator).validateUserOp(userOp, userOpHash);
+        address validator = op.nonce.getValidator();
+        if (op.nonce.isModuleEnableMode()) {
```
Which condition do you think is more likely to happen, this one or the other?

filmakarov replied:
idk. I hope that the permissions won't be one-timers. so switched

Aboudjem reviewed (changes requested):
Can you generate gas report and push it as well?

livingrockrises commented:
@filmakarov can you please resolve conflicts with @Aboudjem 's help now

filmakarov commented:
> LGTM overall, added a few comments
> 
> Can you generate gas report and push it as well?
> 
> `yarn run gas:report`

I added gas report, however it is not representative. As at least two previous PR's (native registry and k1 validator security) haven't provided gas reports. 
So this PR's gas report reflects changes by several commits and it is not easy to find the changes that influenced the gas increase.
